src/daemon/distribution.c

```diff
@@ -0,0 +1,265 @@
+  distribution_t *d = (distribution_t *)calloc(1, sizeof(distribution_t));
+
+  if (d == NULL) {
+    return NULL;
+  }
+
+  d->buckets = bucket_new_exponential(num_buckets, initial_size, factor);
+
+  if (d->buckets == NULL) {
+    free(d);
+    return NULL;
+  }
+
+  d->num_buckets = num_buckets;
+  pthread_mutex_init(&d->mutex, NULL);
+
+  return d;
+}
+
+distribution_t *distribution_new_custom(size_t num_buckets,
+                                        double *custom_buckets_sizes) {
+  distribution_t *d = (distribution_t *)calloc(1, sizeof(distribution_t));
+
+  if (d == NULL) {
+    return NULL;
+  }
+
+  d->buckets = bucket_new_custom(num_buckets, custom_buckets_sizes);
+
+  if (d->buckets == NULL) {
+    free(d);
+    return NULL;
+  }
+
+  d->num_buckets = num_buckets;
+  pthread_mutex_init(&d->mutex, NULL);
+
+  return d;
+}
+
+static void bucket_update(bucket_t *buckets, size_t num_buckets, double gauge) {
+  int ptr = (int)num_buckets - 1;
+
+  while (buckets[ptr].max_boundary > gauge && ptr >= 0) {
+    buckets[ptr].counter++;
+    ptr--;
+  }
+}
+
+void distribution_update(distribution_t *d, double gauge) {
+  if (d == NULL) {
+    errno = EINVAL;
+    return;
+  }
+
+  pthread_mutex_lock(&d->mutex);
+
+  bucket_update(d->buckets, d->num_buckets, gauge);
+
+  d->sum_gauges += gauge;
+  pthread_mutex_unlock(&d->mutex);
+}
+
+static double find_percentile(bucket_t *buckets, size_t num_buckets,
+                              uint64_t quantity) {
+  size_t left = 0;
+  size_t right = num_buckets - 1;
+  size_t middle;
+
+  while (left < right) {
+    middle = (left + right) / 2;
+
+    if (buckets[middle].counter >= quantity) {
+      left = middle;
+    } else {
+      right = middle - 1;
+    }
+  }
+
+  return buckets[left].max_boundary;
+}
+
+double distribution_percentile(distribution_t *d, double percent) {
+  if (d == NULL) {
+    errno = EINVAL;
+    return -1;
+  }
+
+  distribution_t *distribution = distribution_clone(d);
+
+  if (distribution == NULL) {
+    return -1;
+  }
+
+  uint64_t quantity =
+      (percent / 100.0) *
+      distribution->buckets[distribution->num_buckets - 1].counter;
+
+  percent = find_percentile(distribution->buckets, distribution->num_buckets,
+                            quantity);
+
+  distribution_destroy(distribution);
+  return percent;
+}
+
+double distribution_average(distribution_t *d) {
+  if (d == NULL) {
+    errno = EINVAL;
+    return -1;
+  }
+
+  distribution_t *distribution = distribution_clone(d);
+
+  if (distribution == NULL) {
+    return -1;
+  }
+
+  double average =
+      distribution->sum_gauges /
+      (double)distribution->buckets[distribution->num_buckets - 1].counter;
+
+  distribution_destroy(distribution);
+  return average;
+}
+
+distribution_t *distribution_clone(distribution_t *d) {
+  if (d == NULL) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  pthread_mutex_lock(&d->mutex);
+
+  distribution_t *distribution = calloc(1, sizeof(distribution_t));
```

Comment:
I recommend to allocate the memory before locking the mutex. I think we can even assume that `d->num_buckets` doesn't change, i.e. we can read it without holding the lock, which allows us to also allocate `distribution->buckets` without locking.

Reply:
Thanks! I changed that